### DISCUSSION TITLE
howto make "Export program / format: C/C++" only export a subset of the code

### DISCUSSION THREAD

**Discussion by wothke** [NationalSecurityAgency/ghidra#3870] [Q&A]
In my example I am analyzing a MFC based win32 program and I am only interested in one specific library that is used within that program, i.e. 90% of the decompiled code is boilerplate MFC stuff (etc) that I am not interested in at all. I already identified most of the functions that I am actually interested and organized these in a respective package/class structure.

I saw that functions can be individually flagged with user defined "tags" - which can then be used to filter the data export. I would want to tag everything within a specific namespace. But from what I see the "Function Tags" UI only allows to tag functions one by one. It seems excessively silly/annoying  if I had to manually go through hundreds of functions using that dialog just to perform a tagging that I conceptually already had performed via the namespace. 

Is there some bulk-tagging API that I have overlooked? Or some add-on tool that allows to automatically tag based on package name?

**Comment by bdemick** [comment]
`getFunctionAt(address).addTag('EXPORT')` is your (API) friend. Then you can do something like:

```
exp = CppExporter()
opts = [
    Option(CppExporter.FUNCTION_TAG_EXCLUDE, False),
    Option(CppExporter.FUNCTION_TAG_FILTERS, 'EXPORT')
]
exp.setOptions(opts)
exp.setExporterServiceProvider(state.getTool())
exp.export(File("out.c"), currentProgram, None, TaskMonitor.DUMMY)
```

to export the functions that you tagged.

**Reply by wothke** [reply]
> `getFunctionAt(address).addTag('EXPORT')` is your (API) friend. Then you can do something like:

I guess I'd also need to find the APIs needed to integrate a respective script as a 
handler when right-clicking on "namespace nodes", and the APIs needed to then find all the 
functions associated with that namespace, etc. I'd be surprised if I was the first person
looking for that kind of functionality and I had hoped not having to implement it myself. I 
guess I'll have to look into Ghidra's respectice extension APIs myself after all.

> exp = CppExporter()
> opts = [
>     Option(CppExporter.FUNCTION_TAG_EXCLUDE, False),
>     Option(CppExporter.FUNCTION_TAG_FILTERS, 'EXPORT')
> ]
> exp.setOptions(opts)
> exp.setExporterServiceProvider(state.getTool())
> exp.export(File("out.c"), currentProgram, None, TaskMonitor.DUMMY)
> ```
> 
> to export the functions that you tagged.

I guess that is the exact same functionality already available via the 
existing "filter" checkbox of the regular "Export program" functionality.
